client/driver/utils.py

```diff
@@ -0,0 +1,165 @@
+import os
+
+from fabric.api import hide, local, settings, task
+from fabric.api import get as _get, put as _put, run as _run, sudo as _sudo
+
+dconf = None  # pylint: disable=invalid-name
+
+
+def load_driver_conf():
+    # The default config file is 'driver_config.py' but you can use
+    # set the env 'DRIVER_CONFIG' to the path of a different config
+    # file to override it.
+    global dconf  # pylint: disable=global-statement,invalid-name
+    if not dconf:
+        driver_conf = os.environ.get('DRIVER_CONFIG', 'driver_config')
+        if driver_conf.endswith('.py'):
+            driver_conf = driver_conf[:-len('.py')]
+        mod = importlib.import_module(driver_conf)
+        dconf = mod
+
+        # Generate the login string of the host connection
+        if dconf.HOST_CONN == 'local':
+            login_str = 'localhost'
+
+        elif dconf.HOST_CONN == 'remote':
+            if not dconf.LOGIN_HOST:
+                raise ValueError("LOGIN_HOST must be set if HOST_CONN=remote")
+
+            login_str = dconf.LOGIN_HOST
+            if dconf.LOGIN_NAME:
+                login_str = '{}@{}'.format(dconf.LOGIN_NAME, login_str)
+
+            if dconf.LOGIN_PORT:
+                login_str += ':{}'.format(dconf.LOGIN_PORT)
+
+        elif dconf.HOST_CONN == 'docker':
+            if not dconf.CONTAINER_NAME:
+                raise ValueError("CONTAINER_NAME must be set if HOST_CONN=docker")
+            login_str = 'localhost'
+
+        else:
+            raise ValueError(("Invalid HOST_CONN: {}. Valid values are "
+                              "'local', 'remote', or 'docker'.").format(dconf.HOST_CONN))
+        dconf.LOGIN = login_str
+
+    return dconf
+
+
+def parse_bool(value):
+    if not isinstance(value, bool):
+        value = str(value).lower() == 'true'
+    return value
+
+
+@task
+def run(cmd, **kwargs):
+    try:
+        if dconf.HOST_CONN == 'remote':
+            res = _run(cmd, **kwargs)
+        elif dconf.HOST_CONN == 'local':
+            res = local(cmd, capture=True, **kwargs)
+        else:  # docker
+            opts = ''
+            cmdd = cmd
+            if cmd.endswith('&'):
+                cmdd = cmd[:-1].strip()
+                opts = '-d '
+            res = local('docker exec {} -ti {} /bin/bash -c "{}"'.format(
+                opts, dconf.CONTAINER_NAME, cmdd), capture=True, **kwargs)
+    except TypeError as e:
+        err = str(e).strip()
+        if 'unexpected keyword argument' in err:
+            offender = err.rsplit(' ', 1)[-1][1:-1]
+            kwargs.pop(offender)
+            res = run(cmd, **kwargs)
+        else:
+            raise e
+    return res
+
+
+@task
+def sudo(cmd, user=None, **kwargs):
+    if dconf.HOST_CONN == 'remote':
+        res = _sudo(cmd, user=user, **kwargs)
+
+    elif dconf.HOST_CONN == 'local':
+        pre_cmd = 'sudo '
+        if user:
+            pre_cmd += '-u {} '.format(user)
+        res = local(pre_cmd + cmd, capture=True, **kwargs)
```

Comment:
`capture=True` causes postgres restart never completes.
Removing `capture=True` works fine on my machine for the whole pipeline.

Reply:
Fabric's `local` command does not return anything (the stdout, stderr, return_code, etc.) when `capture` is False so I do not want to set it that way in most cases. For now I've added `capture` as an arg to the `sudo` and `run` functions in utils.py. By default it's set to True but I've set it to False in restart_database. If this still doesn't work let me know.